File: backends/fastapi/app/services/auth_service.py
"""鉴权：注册 / 登录 / refresh 轮转 / 登出 / 改密。

关键行为（对应 docs/02-design.md §7、01-requirements.md §4.1）：
  - refresh token 每次换新（rotate）+ family 延续
  - 使用已 revoked 的 token 触发整家族撤销
  - 改密码吊销该用户全部 refresh token
"""

from __future__ import annotations

import uuid
from collections import deque
from datetime import datetime, timedelta, timezone
from time import monotonic

from sqlalchemy import select
from sqlalchemy.exc import IntegrityError
from sqlalchemy.orm import Session

from app.core import errors, security
from app.core.config import settings
from app.models import RefreshToken, User
from app.schemas.user import AuthTokens, UserOut
from app.services.avatar_service import allowed_avatar_ids

# 简单内存限流：email → 最近失败时间戳队列
_login_failures: dict[str, deque[float]] = {}


def _rate_limit_login(email: str) -> None:
    bucket = _login_failures.setdefault(email.lower(), deque())
    now = monotonic()
    # 丢弃 60s 前的记录
    while bucket and now - bucket[0] > 60:
        bucket.popleft()
    if len(bucket) >= settings.login_rate_limit_per_minute:
        raise errors.rate_limited()


def _record_login_failure(email: str) -> None:
    bucket = _login_failures.setdefault(email.lower(), deque())
    bucket.append(monotonic())


def _user_to_dto(u: User) -> UserOut:
    return UserOut(
        id=u.id, email=u.email, nickname=u.nickname, avatarId=u.avatar_id, createdAt=u.created_at
    )


def _issue_token_pair(
    db: Session, user: User, *, family_id: uuid.UUID | None = None
) -> AuthTokens:
    access, access_ttl = security.create_access_token(
        user_id=user.id, nickname=user.nickname, avatar_id=user.avatar_id
    )
    raw_refresh = security.generate_refresh_token()
    token_hash = security.hash_refresh_token(raw_refresh)
    expires_at = datetime.now(timezone.utc) + timedelta(seconds=settings.refresh_token_ttl_seconds)
    row = RefreshToken(
        user_id=user.id,
        token_hash=token_hash,
        family_id=family_id or uuid.uuid4(),
        expires_at=expires_at,
    )
    db.add(row)
    db.flush()
    return AuthTokens(
        accessToken=access,
        refreshToken=raw_refresh,
        accessTokenExpiresIn=access_ttl,
        refreshTokenExpiresIn=settings.refresh_token_ttl_seconds,
        user=_user_to_dto(user),
    )


def register(
    db: Session, *, email: str, password: str, nickname: str, avatar_id: str
) -> AuthTokens:
    if avatar_id not in allowed_avatar_ids():
        raise errors.validation_error("头像 ID 不存在", field="avatarId")

    email_norm = email.lower().strip()
    # 预检提示，精细冲突字段；并发场景下仍由 DB unique 兜底
    if db.execute(select(User).where(User.email == email_norm)).scalar_one_or_none():
        raise errors.conflict("邮箱已被注册", field="email")
    if db.execute(select(User).where(User.nickname == nickname)).scalar_one_or_none():
        raise errors.conflict("昵称已被使用", field="nickname")

    user = User(
        email=email_norm,
        password_hash=security.hash_password(password),
        nickname=nickname,
        avatar_id=avatar_id,
    )
    db.add(user)
    try:
        db.flush()
    except IntegrityError as e:
        db.rollback()
        msg = str(e.orig).lower()
        if "email" in msg:
            raise errors.conflict("邮箱已被注册", field="email") from e
        if "nickname" in msg:
            raise errors.conflict("昵称已被使用", field="nickname") from e
        raise errors.conflict("注册冲突") from e

    tokens = _issue_token_pair(db, user)
    db.commit()
    return tokens


def login(db: Session, *, email: str, password: str) -> AuthTokens:
    email_norm = email.lower().strip()
    _rate_limit_login(email_norm)
    user = db.execute(select(User).where(User.email == email_norm)).scalar_one_or_none()
    if not user or not security.verify_password(password, user.password_hash):
        _record_login_failure(email_norm)
        raise errors.unauthorized("邮箱或密码错误")
    tokens = _issue_token_pair(db, user)
    db.commit()
    return tokens


def refresh(db: Session, *, raw_refresh: str) -> AuthTokens:
    token_hash = security.hash_refresh_token(raw_refresh)
    stmt = select(RefreshToken).where(RefreshToken.token_hash == token_hash)
    if settings.db_driver == "postgres":
        stmt = stmt.with_for_update()
    row = db.execute(stmt).scalar_one_or_none()
    if not row:
        raise errors.unauthorized("refresh token 无效")

    now = datetime.now(timezone.utc)
    if row.expires_at <= now:
        raise errors.unauthorized("refresh token 已过期")

    if row.revoked_at is not None:
        # 重放 → 整族作废
        db.query(RefreshToken).filter(
            RefreshToken.family_id == row.family_id, RefreshToken.revoked_at.is_(None)
        ).update({RefreshToken.revoked_at: now}, synchronize_session=False)
        db.commit()
        raise errors.unauthorized("refresh token 已失效")

    user = db.execute(select(User).where(User.id == row.user_id)).scalar_one_or_none()
    if not user:
        raise errors.unauthorized("用户不存在")

    row.revoked_at = now
    db.flush()
    tokens = _issue_token_pair(db, user, family_id=row.family_id)
    db.commit()
    return tokens


def logout(db: Session, *, raw_refresh: str | None) -> None:
    if not raw_refresh:
        return
    token_hash = security.hash_refresh_token(raw_refresh)
    row = db.execute(
        select(RefreshToken).where(RefreshToken.token_hash == token_hash)
    ).scalar_one_or_none()
    if row and row.revoked_at is None:
        row.revoked_at = datetime.now(timezone.utc)
        db.commit()


def change_password(db: Session, *, user: User, current_password: str, new_password: str) -> None:
    if not security.verify_password(current_password, user.password_hash):
        raise errors.unauthorized("当前密码错误")
    user.password_hash = security.hash_password(new_password)
    # 吊销所有 refresh token
    db.query(RefreshToken).filter(
        RefreshToken.user_id == user.id, RefreshToken.revoked_at.is_(None)
    ).update({RefreshToken.revoked_at: datetime.now(timezone.utc)}, synchronize_session=False)
    db.commit()

File: backends/nest/src/auth/auth.service.ts
import { Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { JwtService } from '@nestjs/jwt';
import { InjectRepository } from '@nestjs/typeorm';
import * as bcrypt from 'bcrypt';
import * as crypto from 'crypto';
import { Repository } from 'typeorm';
import { v4 as uuidv4 } from 'uuid';
import {
  conflict,
  rateLimited,
  unauthorized,
} from '../common/api.exception';
import { AppConfig } from '../config/configuration';
import { Capsule } from '../entities/capsule.entity';
import { RefreshToken } from '../entities/refresh-token.entity';
import { User } from '../entities/user.entity';
import { ALLOWED_AVATAR_IDS } from '../health/avatar.service';
import { validationError } from '../common/api.exception';

export interface UserOut {
  id: string;
  email: string;
  nickname: string;
  avatarId: string;
  createdAt: Date;
}

export interface AuthTokens {
  accessToken: string;
  refreshToken: string;
  accessTokenExpiresIn: number;
  refreshTokenExpiresIn: number;
  user: UserOut;
}

// Simple in-memory rate limiter: email → timestamps of failures within last 60s
const loginFailures = new Map<string, number[]>();

@Injectable()
export class AuthService {
  constructor(
    @InjectRepository(User)
    private readonly userRepo: Repository<User>,
    @InjectRepository(RefreshToken)
    private readonly tokenRepo: Repository<RefreshToken>,
    @InjectRepository(Capsule)
    private readonly capsuleRepo: Repository<Capsule>,
    private readonly jwtService: JwtService,
    private readonly config: ConfigService<AppConfig>,
  ) {}

  private hashPassword(plain: string): string {
    const rounds = this.config.get('bcryptRounds', { infer: true }) ?? 12;
    return bcrypt.hashSync(plain, rounds);
  }

  private verifyPassword(plain: string, hashed: string): boolean {
    try {
      return bcrypt.compareSync(plain, hashed);
    } catch {
      return false;
    }
  }

  private generateRefreshToken(): string {
    return crypto.randomBytes(32).toString('base64url');
  }

  private hashRefreshToken(raw: string): string {
    return crypto.createHash('sha256').update(raw).digest('hex');
  }

  private checkRateLimit(email: string): void {
    const limit = this.config.get('loginRateLimitPerMinute', { infer: true }) ?? 10;
    const now = Date.now();
    const failures = (loginFailures.get(email) ?? []).filter((t) => now - t < 60_000);
    loginFailures.set(email, failures);
    if (failures.length >= limit) throw rateLimited();
  }

  private recordFailure(email: string): void {
    const failures = loginFailures.get(email) ?? [];
    failures.push(Date.now());
    loginFailures.set(email, failures);
  }

  private toUserOut(u: User): UserOut {
    return { id: u.id, email: u.email, nickname: u.nickname, avatarId: u.avatarId, createdAt: u.createdAt };
  }

  private issueTokenPair(user: User, familyId?: string): Omit<AuthTokens, 'user'> & { tokenRow: Partial<RefreshToken> } {
    const accessTtl = this.config.get('accessTokenTtlSeconds', { infer: true }) ?? 900;
    const refreshTtl = this.config.get('refreshTokenTtlSeconds', { infer: true }) ?? 2592000;

    const accessToken = this.jwtService.sign(
      { sub: user.id, nickname: user.nickname, avatarId: user.avatarId },
      { expiresIn: accessTtl },
    );

    const rawRefresh = this.generateRefreshToken();
    const tokenHash = this.hashRefreshToken(rawRefresh);
    const now = new Date();
    const expiresAt = new Date(now.getTime() + refreshTtl * 1000);

    const tokenRow: Partial<RefreshToken> = {
      id: uuidv4(),
      userId: user.id,
      tokenHash,
      familyId: familyId ?? uuidv4(),
      expiresAt,
      createdAt: now,
      revokedAt: null,
    };

    return {
      accessToken,
      refreshToken: rawRefresh,
      accessTokenExpiresIn: accessTtl,
      refreshTokenExpiresIn: refreshTtl,
      tokenRow,
    };
  }

  async register(dto: {
    email: string;
    password: string;
    nickname: string;
    avatarId: string;
  }): Promise<AuthTokens> {
    if (!ALLOWED_AVATAR_IDS.has(dto.avatarId)) {
      throw validationError('头像 ID 不存在', 'avatarId');
    }

    const emailNorm = dto.email.toLowerCase().trim();

    const existingEmail = await this.userRepo.findOne({ where: { email: emailNorm } });
    if (existingEmail) throw conflict('邮箱已被注册', 'email');

    const existingNick = await this.userRepo.findOne({ where: { nickname: dto.nickname } });
    if (existingNick) throw conflict('昵称已被使用', 'nickname');

    const now = new Date();
    const user: User = this.userRepo.create({
      id: uuidv4(),
      email: emailNorm,
      passwordHash: this.hashPassword(dto.password),
      nickname: dto.nickname,
      avatarId: dto.avatarId,
      createdAt: now,
      updatedAt: now,
    });

    try {
      await this.userRepo.save(user);
    } catch (e: any) {
      const msg = (e?.message ?? '').toLowerCase();
      if (msg.includes('email')) throw conflict('邮箱已被注册', 'email');
      if (msg.includes('nickname')) throw conflict('昵称已被使用', 'nickname');
      throw conflict('注册冲突');
    }

    const { tokenRow, ...tokens } = this.issueTokenPair(user);
    await this.tokenRepo.save(this.tokenRepo.create(tokenRow));
    return { ...tokens, user: this.toUserOut(user) };
  }

  async login(dto: { email: string; password: string }): Promise<AuthTokens> {
    const emailNorm = dto.email.toLowerCase().trim();
    this.checkRateLimit(emailNorm);

    const user = await this.userRepo.findOne({ where: { email: emailNorm } });
    if (!user || !this.verifyPassword(dto.password, user.passwordHash)) {
      this.recordFailure(emailNorm);
      throw unauthorized('邮箱或密码错误');
    }

    const { tokenRow, ...tokens } = this.issueTokenPair(user);
    await this.tokenRepo.save(this.tokenRepo.create(tokenRow));
    return { ...tokens, user: this.toUserOut(user) };
  }

  async refresh(rawRefresh: string): Promise<AuthTokens> {
    // 教学注释：生产环境应用行锁（SELECT ... FOR UPDATE）或原子操作
    // （UPDATE ... WHERE revoked_at IS NULL RETURNING *）消除并发 refresh 的双发窗口。
    const tokenHash = this.hashRefreshToken(rawRefresh);
    const row = await this.tokenRepo.findOne({ where: { tokenHash } });
    if (!row) throw unauthorized('refresh token 无效');

    const now = new Date();
    if (row.expiresAt <= now) throw unauthorized('refresh token 已过期');

    if (row.revokedAt !== null) {
      // Reuse detected — revoke entire family
      await this.tokenRepo
        .createQueryBuilder()
        .update()
        .set({ revokedAt: now })
        .where('family_id = :fid AND revoked_at IS NULL', { fid: row.familyId })
        .execute();
      throw unauthorized('refresh token 已失效');
    }

    const user = await this.userRepo.findOne({ where: { id: row.userId } });
    if (!user) throw unauthorized('用户不存在');

    row.revokedAt = now;
    await this.tokenRepo.save(row);

    const { tokenRow, ...tokens } = this.issueTokenPair(user, row.familyId);
    await this.tokenRepo.save(this.tokenRepo.create(tokenRow));
    return { ...tokens, user: this.toUserOut(user) };
  }

  async logout(rawRefresh: string | undefined): Promise<void> {
    if (!rawRefresh) return;
    const tokenHash = this.hashRefreshToken(rawRefresh);
    const row = await this.tokenRepo.findOne({ where: { tokenHash } });
    if (row && row.revokedAt === null) {
      row.revokedAt = new Date();
      await this.tokenRepo.save(row);
    }
  }

  async changePassword(
    user: User,
    dto: { currentPassword: string; newPassword: string },
  ): Promise<void> {
    if (!this.verifyPassword(dto.currentPassword, user.passwordHash)) {
      throw unauthorized('当前密码错误');
    }
    user.passwordHash = this.hashPassword(dto.newPassword);
    user.updatedAt = new Date();
    await this.userRepo.save(user);
    // Revoke all refresh tokens
    await this.tokenRepo
      .createQueryBuilder()
      .update()
      .set({ revokedAt: new Date() })
      .where('user_id = :uid AND revoked_at IS NULL', { uid: user.id })
      .execute();
  }

  async findUserById(id: string): Promise<User | null> {
    return this.userRepo.findOne({ where: { id } });
  }
}
